Eq 1113 refactor navigation

dwyeradam opened the pull request:
### What is the context of this PR?
Currently the Navigation has been overloaded into the groups in the schema. This was done based on the assumption there was a one-to-one mapping between a group and a navigation section. As the groups have been used to control flow and repeating, this is not the case. 

We should re-implement navigation to be defined outside of the group structure in the schema

### How to review 
- It is possible to identify which groups need to be completed for a navigation item to be marked as complete (get a tick) - multiple groups are allowed.
- What constitutes a group being complete is defined by the blocks within it and the route taken (all routed blocks must be completed).
- The group to jump to when using a navigation link can be specified. Only one group can be given.
- The title of the navigation link (section) can be customised.

jonnyshaw89 commented on app/questionnaire/navigation.py:
```diff
@@ -31,86 +31,96 @@ def build_navigation(self, current_group_id, current_group_instance):
-    def _should_skip_group(self, current_group_instance, group):
-
-        skip_group = False
+    def _get_last_sect_group(self, section, current_group_instance):
```
Is there any need for the abbreviation in this method?

dwyeradam replied:
Nope so now fully formed as _get_last_section_group.

jonnyshaw89 replied:
Is it?

dwyeradam replied:
Yes he says sheepishly. Seems I lost a couple of changes during the conflict resolution. Done now.

dcdarrell9 reviewed (commented):
Looks good 👍

dwyeradam commented:
Adding functional tests to check for completed status.

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,133 @@ def build_navigation(self, current_group_id, current_group_instance):
-            last_block_location = Location(group['id'], 0, last_block_in_group['id'])
-            can_get_to_summary = is_summary_or_confirm_group and self.routing_path and \
-                last_block_location in self.routing_path
+            if first_block_id in ('summary', 'confirmation') \
```
This needs to use the `is_summary_or_confirmation` method in `SchemaHelper` so that we're comparing against block types rather than ids

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,133 @@ def build_navigation(self, current_group_id, current_group_instance):
-        for group in filter(lambda x: 'hide_in_navigation' not in x, self.survey_json['groups']):
-            logger.debug("building frontend navigation", group_id=group['id'])
+        for section in navigation_block['sections']:
+            first_section_group_id = self._get_first_non_skipped_group_in_section(section)
```
I think it would be clearer if this was named `first_group_id_in_section` or possibly dropping the section and making it `first_group_id`

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,133 @@ def build_navigation(self, current_group_id, current_group_instance):
 
-            is_summary_or_confirm_group = last_block_in_group and \
-                SchemaHelper.is_summary_or_confirmation(last_block_in_group)
+            last_section_group_id = self._get_last_non_skipped_group_in_section(section)
```
Same as comment above, this could be `last_group_id`
Also, as this variable is only used if the block is a summary or confirmation, we should move it under that conditional check, something like:
```
if first_block_id in ('summary', 'confirmation'):
    last_section_group_id = self._get_last_non_skipped_group_in_section(section)
    if self._can_reach_summary_confirmation(last_section_group_id, last_block_id) is False:
        continue
```

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,133 @@ def build_navigation(self, current_group_id, current_group_instance):
+            if not self._should_skip_group(group_id):
+                return group_id
+
+        return section['group_order'][0]
```
Why does this return the first group id even if it's skipped?

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,133 @@ def build_navigation(self, current_group_id, current_group_instance):
+
+        return skip_nav
+
+    def _get_last_block_id(self, last_section_group_id, current_group_instance):
```
`last_section_group_id` should be `group_id` - the method isn't concerned with where the group id came from

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,133 @@ def build_navigation(self, current_group_id, current_group_instance):
-        return self._generate_item(group['title'], is_completed, first_location, default_highlight)
+    def _build_single_repeating_navigation(self, section, current_group_id, first_location,
+                                           completed_id, no_of_repeats):
+        default_highlight = current_group_id in section['group_order']
```
Not sure why this was called `default_highlight` previously, it should be renamed something like `should_highlight`

jonnyshaw89 commented:
@dwyeradam can you rebase this branch so it had the latest changes from master?

jonnyshaw89 commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,145 @@ def build_navigation(self, current_group_id, current_group_instance):
+            for block in reversed(group['blocks']):
+                skip_cons = block.get('skip_conditions')
+                if evaluate_skip_conditions(skip_cons, self.metadata, self.answer_store, group_instance) is False:
+                    last_group_block_id = {'group_id': group_id, 'block_id': block['id']}
```
could you not just return here?

jonnyshaw89 commented:
One issue i found was that if i complete a block but then go back and change an answer that routes. The section is still marked as complete when the newly exposed block has not been completed.
The result is that when reaching the end of the survey i am redirected back to the un-completed block.
Discussed with @dwyeradam and a separate card will be raised to address this.

dwyeradam commented:
Just to expand and clarify on  @jonnyshaw89 comment above. This is where the new routing path shares the same end point as a previously completed path. This is because the code is only looking at the completed id of the last answer which in this instance would remain completed. 
Changes to answers where the end point is changed as a result will remove the completed mark.

jonnyshaw89 commented on tests/new_functional/spec/navigation.spec.js:
```diff
@@ -0,0 +1,156 @@
+const SkipPaymentPage = require('../pages/surveys/navigation/skip-payment.page.js');
+const SummaryPage = require('../pages/surveys/navigation/summary.page.js');
+
+describe('Navigation', function() {
```
Were we going to add tests for the section 'ticks'?

dwyeradam replied:
Yes if I can figure out how. Had some difficulty getting it to recognise them.

yoganandkunche commented:
@dwyeradam: As discussed earlier if you can add at least one more question to that group which has repeating questions will allow me to cover more scenarios where I can confirm that the last question after the repeating questions is considered as the trigger to set the group as completed (Green Tick).

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,141 @@ def build_navigation(self, current_group_id, current_group_instance):
+            group = SchemaHelper.get_group(self.survey_json, group_id)
+            for block in reversed(group['blocks']):
+                skip_cons = block.get('skip_conditions')
+                if evaluate_skip_conditions(skip_cons, self.metadata, self.answer_store, group_instance) is False:
```
`group_instance` is the current group instance the user is on when `build_navigation` is called, and doesn't change each time `_get_last_group_and_block_id` is called. I think for now we should remove this argument and let `evaluate_skip_conditions` use the default of `0`.  This will mean that for every repeat of a group the completed block will be the same, but we should tackle that problem when we look at using the routing path to determine if a section is complete.

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,143 @@ def build_navigation(self, current_group_id, current_group_instance):
+         block is obtained.
+         Assumption for repeating groups is that the
+         final block is
+         constant across group instances
```
Minor point, but not sure what's going on with the new lines in this documentation.

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,143 @@ def build_navigation(self, current_group_id, current_group_instance):
-        return skip_group
+            skip_nav = evaluate_skip_conditions(skip_conditions, self.metadata, self.answer_store)
+
+        return skip_nav
```
`skip_nav` should be renamed something like `skip_group` as this method doesn't know anything about navigation

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,143 @@ def build_navigation(self, current_group_id, current_group_instance):
 
-            first_location = Location(group['id'], 0, group['blocks'][0]['id'])
-            last_block_in_group = SchemaHelper.get_last_block_in_group(group)
+            last_group_block_id = self._get_last_group_and_block_id(non_skipped_groups)
```
I think it would be clearer if we replaced references to `last` with `completion`, so we'd have `completion_group_and_block_id`, `_get_completion_group_and_block_id`, and further on `completion_group_id` and `completion_block_id`

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,143 @@ def build_navigation(self, current_group_id, current_group_instance):
+        for group_id in reversed(group_list):
+            group = SchemaHelper.get_group(self.survey_json, group_id)
+            for block in reversed(group['blocks']):
+                skip_cons = block.get('skip_conditions')
```
`skip_cons` shouldn't be abbreviated, it should be `skip_conditions`

ajmaddaford commented on app/questionnaire/navigation.py:
```diff
@@ -31,110 +31,143 @@ def build_navigation(self, current_group_id, current_group_instance):
+         final block is
+         constant across group instances
+
+        :param group_list: List of groups to check
```
It might be clearer to call this `group_ids` rather than `group_list`

yoganandkunche commented:
@dwyeradam Error 500 on final submission screen needs fixing

ajmaddaford commented on tests/new_functional/config.js:
```diff
@@ -38,6 +39,9 @@ let config = {
     // Allow chaining of browser promises with chai assertions
     chaiAsPromised.transferPromiseness = browser.transferPromiseness;
+
+    browser.addCommand("checkCompletedTick", helpers.checkCompletedTick);
```
As discussed, I think we should look at how we can implement this as a helper method in the test rather than as a browser command.

ajmaddaford commented on tests/new_functional/spec/navigation.spec.js:
```diff
@@ -0,0 +1,204 @@
+          return browser.elements('.question__answers > div').then(function (e) {
+            return e.value.length === 1;
+          });
+        }, 2000, 'Person not removed in time')
```
Not sure why we need to wait for the person to be removed before submitting. Is this an issue we need to look into as it could affect users completing the survey?

dwyeradam replied:
Potentially yes but in reality I doubt it will really happen but we should look into it. In practice you have to be very quick with the mouse but if you are quick enough then the submit can preempt the remove.

ajmaddaford commented on tests/new_functional/helpers.js:
```diff
@@ -81,6 +81,39 @@ function isViewSectionsVisible() {
   return browser.isExisting(viewSectionsLink) && browser.isVisibleWithinViewport(viewSectionsLink);
 }
 
+function checkCompletedTick(link_name) {
```
I think this method can be removed now it has been replaced with `isSectionComplete`?

yoganandkunche reviewed (approved):
LGTM 👍🏼

dwyeradam committed:
EQ 1113 Refactor Navigation

Navigation defined in "navigation" and is independent from group structure.
Navigation is made up of sections defining the link names and groups to be included in a particular navigation link.
Each section has it's own title which will appear as the link name. Title from answers will take answer values for the link name.
Each section can be made up of one or more groups.
Navigation links will go to the first group in the section declaration.
Navigation will be marked as complete when all groups within a section are completed.
Functional tests expanded to include helpers for navigation which can be used across all tests. These can select links and check completed status.

shahi645 reviewed (dismissed):
testing fourth wall
